Hazelcast.Net/Hazelcast.Client.Protocol/ClientMessage.cs

```diff
@@ -148,14 +149,24 @@ public virtual bool IsRetryable()
 
         public virtual bool ReadFrom(ByteBuffer source)
         {
-            if (Index() == 0)
+            var frameLength = 0;
+            if (Buffer == null)
             {
-                InitFrameSize(source);
-            }
-            while (Index() >= Bits.IntSizeInBytes && source.HasRemaining() && !IsComplete())
-            {
-                Accumulate(source, GetFrameLength() - Index());
+                //init internal buffer
+                var remaining = source.Remaining();
+                if (remaining < Bits.IntSizeInBytes) {
+                    //we don't have even the frame length ready
+                    return false;
+                }
+                frameLength = Bits.ReadIntL(source.Array(), source.Position);
+                if (frameLength < HeaderSize)
+                {
+                    throw new InvalidDataException("Client message frame length cannot be smaller than header size.");
+                }
+                Wrap(new SafeBuffer(new byte[frameLength]), 0);
             }
+            frameLength = frameLength > 0 ? frameLength : GetFrameLength();
+            Accumulate(source, frameLength - Index());
```

Comment:
Should we also copy the first IntSizeInBytes bytes (the bytes that indicates the frame length) into the Buffer?